README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
pointdexter <img src="man/figures/logo.png" align="right" alt="" width="140" />
===============================================================================

<!-- badges: start -->
[![CRAN status](https://www.r-pkg.org/badges/version/pointdexter)](https://cran.r-project.org/package=pointdexter) [![CRAN checks](https://cranchecks.info/badges/worst/pointdexter)](https://cran.r-project.org/web/checks/check_results_pointdexter.html) [![CRAN Downloads](http://cranlogs.r-pkg.org/badges/pointdexter?color=brightgreen)](http://www.r-pkg.org/pkg/pointdexter) [![Rdoc](http://www.rdocumentation.org/badges/version/pointdexter)](http://www.rdocumentation.org/packages/pointdexter) [![Travis build status](https://travis-ci.org/cenuno/pointdexter.svg?branch=master)](https://travis-ci.org/cenuno/pointdexter) <!-- badges: end -->

The `pointdexter` package labels longitudinal and latitudinal coordinates located inside a polygon.

Description
-----------

For a singular polygon, the label for each coordinate pair is a logical vector of TRUE and FALSE values. For multiple polygons, the label for each coordinate pair is a character vector based on the names of each polygon.

Spatial Packages
----------------

The package is designed to work with both [sf](https://r-spatial.github.io/sf/) and [SpatialPolygonsDataFrame](https://www.rdocumentation.org/packages/sp/versions/1.2-5/topics/SpatialPolygonsDataFrame-class) objects.

By default, `pointdexter` only installs and loads the [`sp`](https://www.rdocumentation.org/packages/sp/versions/1.3-1) and [`splancs`](https://www.rdocumentation.org/packages/splancs/versions/2.01-40) packages.

However, `pointdexter` also works with the [`sf`](https://www.rdocumentation.org/packages/sf/versions/0.7-2) package. The following is `sf` installation advice from [Matt Herman](https://nycgeo.mattherman.info/):

> Depending on your operating system and available libraries, `sf` can be tricky to install the first time. The [`sf` website](https://r-spatial.github.io/sf/index.html#installing) is a good place to start if you're having trouble. If you're using macOS, [this is a good guide](https://medium.com/@jinwujour/mapping-with-r-on-mac-installation-8c8ef997c6c2) to installing the required libraries.

Installation
------------

The latest stable release version can be installed from [CRAN](https://cran.r-project.org/package=pointdexter):

``` r
install.packages("pointdexter")
```

Development versions can be installed from GitHub:

``` r
# note: by default, the development version is the master branch;
#       however, that can be changed by changing the 
#       value in the 'ref' argument
remotes::install_github("cenuno/pointdexter")
```

Usage
-----

### Build-in Data

`pointdexter` comes with built-in point and polygon data - entirely due to the awesome and accessible [Chicago Data Portal](https://data.cityofchicago.org/) - to help you label points in polygons:

-   Chicago Public Schools (CPS) - [School Profile Information, School Year (SY) 2018-2019](https://data.cityofchicago.org/Education/Chicago-Public-Schools-School-Profile-Information-/kh4r-387c)
    -   as a data frame
-   City of Chicago [boundary](https://data.cityofchicago.org/Facilities-Geographic-Boundaries/Boundaries-City/ewy2-6yfk)
    -   as a SpatialPolygonsDataFrame; and
    -   as a simple feature
-   Chicago's [77 community areas](https://data.cityofchicago.org/Facilities-Geographic-Boundaries/Boundaries-Community-Areas-current-/cauq-8yn6)
    -   as a SpatialPolygonsDataFrame; and
    -   as a simple feature
-   Chicago's [2010 census tracts](https://data.cityofchicago.org/Facilities-Geographic-Boundaries/Boundaries-Census-Tracts-2010/5jrd-6zik)
    -   as a SpatialPolygonsDataFrame; and
    -   as a simple feature

### Example

Using built-in data, down below is an example of how to use `pointdexter` to label coordinate pairs with their appropriate polygon:

``` r
# load necessary packages ----
library(pointdexter) # label coordinate pairs in polygons
library(sp)          # classes and methods for spatial data
library(knitr)       # general purpose package for dynamic report generation

# load necessary data ----
data("cps_sy1819")
data("community_areas_spdf")

# create list of coordinate pair matrices for each community area ----
community.area.boundaries <-
  GetPolygonBoundaries(my.polygon = community_areas_spdf
                       , labels = community_areas_spdf$community)

# identify the community that each cps school lies in ----
cps_sy1819$community <-
  LabelPointsWithinPolygons(lng = cps_sy1819$school_longitude
                            , lat = cps_sy1819$school_latitude
                            , polygon.boundaries = community.area.boundaries)

# store relevant columns ----
relevant.columns <-
  c("school_id", "short_name"
    , "school_longitude", "school_latitude", "community")

# show first few records ----
kable(head(cps_sy1819[, relevant.columns], n = 10))
```

| school\_id | short\_name                  |  school\_longitude|  school\_latitude| community       |
|:-----------|:-----------------------------|------------------:|-----------------:|:----------------|
| 609760     | CARVER MILITARY HS           |          -87.59062|          41.65629| RIVERDALE       |
| 609780     | MARINE LEADERSHIP AT AMES HS |          -87.72174|          41.91604| LOGAN SQUARE    |
| 610304     | PHOENIX MILITARY HS          |          -87.68696|          41.87912| NEAR WEST SIDE  |
| 610513     | AIR FORCE HS                 |          -87.63276|          41.82814| ARMOUR SQUARE   |
| 610390     | RICKOVER MILITARY HS         |          -87.66579|          41.98902| EDGEWATER       |
| 609754     | CHICAGO MILITARY HS          |          -87.61922|          41.83055| DOUGLAS         |
| 610132     | POE                          |          -87.60642|          41.70263| PULLMAN         |
| 610177     | SKINNER                      |          -87.65942|          41.87910| NEAR WEST SIDE  |
| 610534     | SKINNER NORTH                |          -87.64484|          41.90511| NEAR NORTH SIDE |
| 610066     | MCDADE                       |          -87.61921|          41.73455| CHATHAM         |

``` r
# plot only CPS high schools in the the Austin, Rogers Park, and West Elsdon community areas ----

# create filter condition ----
filter.condition <- 
  which(cps_sy1819$is_high_school &
          cps_sy1819$community %in% c("AUSTIN"
                                      , "ROGERS PARK"
                                      , "WEST ELSDON"))

# filter cps records to those that matched our condition ---
df <- cps_sy1819[filter.condition, ]

# note: clear plot space
par(mar = c(0, 0, 1, 0))
plot(x = community_areas_spdf
     , main = "CPS High Schools in Austin, Rogers Park, and West Elsdon communities, SY1819"
     , cex.main = 0.75
     , col = "gray85"
     , border = "dodgerblue4")
points(x = df$school_longitude
       , y = df$school_latitude
       , pch = 19
       , col = rgb(red = 212, green = 69, blue = 0
                   , alpha = 90
                   , maxColorValue = 255)
       , cex = 1)
```

<img src="man/figures/README-visualize-1.png" width="100%" />

``` r

# store relevant columns ----
relevant.columns <-
  c("school_id", "short_name", "is_high_school"
    , "school_longitude", "school_latitude", "community")

# print table of those schools that met the condition ----
kable(df[order(df$community), relevant.columns], row.names = FALSE)
```

| school\_id | short\_name                   | is\_high\_school |  school\_longitude|  school\_latitude| community   |
|:-----------|:------------------------------|:-----------------|------------------:|-----------------:|:------------|
| 400127     | YCCS - AUSTIN CAREER          | TRUE             |          -87.76022|          41.89498| AUSTIN      |
| 400123     | YCCS - SCHOLASTIC ACHIEVEMENT | TRUE             |          -87.74254|          41.88045| AUSTIN      |
| 400144     | YCCS - WESTSIDE HOLISTIC      | TRUE             |          -87.74881|          41.90224| AUSTIN      |
| 610245     | DOUGLASS HS                   | TRUE             |          -87.76767|          41.89037| AUSTIN      |
| 610244     | CLARK HS                      | TRUE             |          -87.75333|          41.87288| AUSTIN      |
| 610518     | AUSTIN CCA HS                 | TRUE             |          -87.76192|          41.88599| AUSTIN      |
| 400035     | CHICAGO MATH & SCIENCE HS     | TRUE             |          -87.67482|          42.01303| ROGERS PARK |
| 609733     | SULLIVAN HS                   | TRUE             |          -87.66919|          42.00269| ROGERS PARK |
| 609694     | HANCOCK HS                    | TRUE             |          -87.72447|          41.79146| WEST ELSDON |

Resources
---------

After you've installed the package, be sure to view the help files that introduce you to `pointdexter`'s two functions:

1.  [`?pointdexter::GetPolygonBoundaries()`](https://cenuno.github.io/pointdexter/reference/GetPolygonBoundaries.html); and
2.  [`?pointdexter::LabelPointsWithinPolygons()`](https://cenuno.github.io/pointdexter/reference/LabelPointsWithinPolygons.html).

Feedback
--------

### Cite

If you use `pointdexter` for any analysis, I would love to hear about it! You can also cite the package according to `citation("pointdexter")`.

### Contribute

[Issues](https://github.com/cenuno/pointdexter/issues) and [pull requests](https://github.com/cenuno/pointdexter/pulls) are welcome anytime!

Code of Conduct
---------------

Please note that this project is released with a [Contributor Code of Conduct](.github/CODE_OF_CONDUCT.md). By participating in this project you agree to abide by its terms.


NEWS.md
# pointdexter 0.1.1

* Added 2010 Chicago census tracts as built-in data sets as both `sf` and `SpatialPolygonDataFrame` objects
* Fixed an error where `LabelPointsWithinPolygons()` failed when no point(s) lay within a polygon. 
    + By using smaller geographies (i.e. [census tracts](https://www.census.gov/geo/reference/gtc/gtc_ct.html)) rather than larger ones (i.e. [community areas](http://www.encyclopedia.chicagohistory.org/pages/319.html)), `pointdexter`'s 0.1.0 method of labeling did not account for instances where a polygon has no points laying inside of it.
    + Resolved by storing the results of [`splancs::inpip()`](https://www.rdocumentation.org/packages/splancs/versions/2.01-40/topics/inpip) for each polygon as a named integer vector in a list rather than in a data frame.

# pointdexter 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/data.R
#' CPS - School Profile Information, SY 2018-2019
#'
#' School profile information for all schools in the Chicago Public School (CPS)
#' district for the school year (SY) 2018-2019.
#'
#' @format A data frame of 660 observations - one record per school - and 81 variables
#' @source City of Chicago Data Portal \url{https://data.cityofchicago.org/Education/Chicago-Public-Schools-School-Profile-Information-/kh4r-387c}
#' @name cps_sy1819
NULL


#'
#' A SpatialPolygonsDataFrame of the city boundary of Chicago
#'
#' A SpatialPolygonsDataFrame of the city boundary of Chicago
#' that can only be used with the sp and rgdal packages.
#'
#' @format A SpatialPolygonsDataFrame with 1 observation and 4 variables
#' @source City of Chicago Data Portal \url{https://data.cityofchicago.org/Facilities-Geographic-Boundaries/Boundaries-City/ewy2-6yfk}
#' @name city_boundary_spdf
NULL


#' A simple feature containing the city boundary of Chicago.
#'
#' A simple feature containing the city boundary of Chicago
#' that can only be used with the sf package
#'
#' @format A simple feature with 1 observation and 5 variables
#' @source City of Chicago Data Portal \url{https://data.cityofchicago.org/Facilities-Geographic-Boundaries/Boundaries-City/ewy2-6yfk}
#' @name city_boundary_sf
NULL


#' A SpatialPolygonsDataFrame of Chicago's 77 community areas
#'
#' 77 community area boundaries in Chicago that only be used with
#' the sp and rgdal packages
#'
#' @format A SpatialPolygonsDataFrame with 77 observations and 2 variables
#' @source City of Chicago Data Portal \url{https://data.cityofchicago.org/Facilities-Geographic-Boundaries/Boundaries-Community-Areas-current-/cauq-8yn6}
#' @name community_areas_spdf
NULL


#'
#' A simple feature of Chicago's 77 community areas
#'
#' 77 community area boundaries in Chicago that only be used with
#' the sf package
#'
#' @format A simple feature with 77 observations and 3 variables
#' @source City of Chicago Data Portal \url{https://data.cityofchicago.org/Facilities-Geographic-Boundaries/Boundaries-Community-Areas-current-/cauq-8yn6}
#' @name community_areas_sf
NULL



R/pointdexter.R
#' pointdexter: labels longitudinal and latitudinal coordinates located inside a polygon
#'
#' The pointdexter package labels coordinate pairs located inside a polygon.
#' For a singular polygon, the label is a logical vector of TRUE and FALSE values.
#' For multiple polygons, the label is a character vector based on the names of each polygon.
#' The package is designed to work with both \href{https://r-spatial.github.io/sf/}{\code{sf}} and
#'  \href{https://www.rdocumentation.org/packages/sp/versions/1.2-5/topics/SpatialPolygonsDataFrame-class}{\code{SpatialPolygonsDataFrame}} objects.
#'
#' To learn more about pointdexter, start with the vignette:
#' \code{browseVignettes(package = "pointdexter")}
#'
#' @section pointdexter functions:
#' \describe{
#' \item{\code{GetPolygonBoundaries()}}{obtains the boundaries of the polygon(s)}
#' \item{\code{LabelPointsWithinPolygons()}}{labels points located a polygon}
#' }
#'
#' @docType package
#' @name pointdexter
NULL


R/01_get_polygon_boundaries.R
#
# Author:   Cristian E. Nuno
# Purpose:  Retrieve coordinate pairs from each polygon in a spatial polygon data frame
# Date:     December 28, 2018
#

# create GetPolygonBoundaries() function -----
#' @export
GetPolygonBoundaries <- function(my.polygon
                                 , labels = NULL) {
  # Obtains the boundaries of the polygon(s)
  #
  # Input: ----
  # my.polygon: either a spatial polygons data frame or a sf object
  # labels: a character vector of polygon boundary labels used to name each matrix

  # Output: ----
  # If my.polygon is of length 1, a coordinate pair matrix;
  # else a list of labeled matrices,
  # with each matrix representing the coordinate pairs
  # that make the boundary of each particular polygon.
  # Each matrix containing a unique record for every unique coordinate pair
  # longitude is the first column and latitude is the second column

  # ensure my.polygon is correct class -----
  if (!any(class(my.polygon) %in%
          c("sf", "SpatialPolygonsDataFrame"))) {
    stop("my.polygon is of class "
          , class(my.polygon)[1]
          , ". Please ensure it is of class SpatialPolygonsDataFrame or sf.\n\nSee ?sp::`SpatialPolygonsDataFrame-class` or vignette('sf1', 'sf') for more help.")
  }

  # ensure that if my.polygon is an sf object, that the sf package is installed -----
  if (class(my.polygon)[1] == "sf" &
      requireNamespace("sf", quietly = TRUE) == FALSE) {
    stop("my.polygon is of class sf yet you do not have the sf package.\n\nEither install the sf package - install.packages('sf') - or input a SpatialPolygonsDataFrame object into my.polygon.")
  }

  # ensure that labels is a character vector ----
  if (!is.null(labels) & !is.character(labels)) {
    stop("labels must be a character vector.")
  }

  # print message that labels won't be used with one polygon ----
  if (nrow(my.polygon) == 1 & !is.null(labels)) {
    message("No label is assigned to my.polygon of length 1.")
  }

  # ensure that labels is same number of rows as my.polygon ----
  if (!is.null(labels) &
      length(labels) != nrow(my.polygon)) {
    stop("The number of elements in labels must equal the number of rows in my.polygon.")
  }

  # if there is only 1 polygon, retrieve its coordinate pair matrix ----
  if (nrow(my.polygon) == 1) {
    if (class(my.polygon)[1] == "SpatialPolygonsDataFrame") {
      polygon.boundary <-
        lapply(X = my.polygon@polygons[[1]]@Polygons
               , FUN = function(i) i@coords)
    } else {
      polygon.boundary <-
        lapply(X = sf::st_boundary(my.polygon$geometry)[[1]]
               , FUN = function(i) i)
    }
    # to account for holes in a polygon, only extract the coordinate pair matrix with the most points
    max.poly <- which.max(sapply(X = polygon.boundary, FUN = nrow))

    return(polygon.boundary[[max.poly]])
    # if this is more than 1 polygon, retrieve each polygon's coordinate pair matrix ----
  } else {
    if (class(my.polygon)[1] == "SpatialPolygonsDataFrame") {
      polygon.boundaries <-
        lapply(X = my.polygon@polygons
               , FUN = function(i)
                 i@Polygons[[1]]@coords)
    } else {
      polygon.boundaries <-
        lapply(X = sf::st_boundary(my.polygon$geometry)
               , FUN = function(i) i[[1]])
    }
    if (!is.null(labels)) {
      names(polygon.boundaries) <- labels
    }
    return(polygon.boundaries)
  }
} # end of GetPolygonBoundaries() function

# end of script #


data-raw/01_import_raw_data.R
#
# Author:   Cristian E. Nuno
# Purpose:  Import raw data sets
# Date:     March 14, 2019
#

# load necessary data ----

# 2010 chicago census tracts permalink:     https://data.cityofchicago.org/Facilities-Geographic-Boundaries/Boundaries-Census-Tracts-2010/5jrd-6zik
# community areas' permalink:               https://data.cityofchicago.org/Facilities-Geographic-Boundaries/Boundaries-Community-Areas-current-/cauq-8yn6
# city boundary permalink:                  https://data.cityofchicago.org/Facilities-Geographic-Boundaries/Boundaries-City/ewy2-6yfk
# SY1819 chicago public schools permalink:  https://data.cityofchicago.org/Education/Chicago-Public-Schools-School-Profile-Information-/kh4r-387c

chicago.data <-
  list("chicago_2010_census_tracts.geojson" = "https://data.cityofchicago.org/api/geospatial/5jrd-6zik?method=export&format=GeoJSON"
    , "chicago_community_area_boundaries.geojson" = "https://data.cityofchicago.org/api/geospatial/cauq-8yn6?method=export&format=GeoJSON"
    , "chicago_city_boundary.geojson" = "https://data.cityofchicago.org/api/geospatial/ewy2-6yfk?method=export&format=GeoJSON"
    , "chicago_public_schools_sy1819.csv" = "https://data.cityofchicago.org/api/views/kh4r-387c/rows.csv?accessType=DOWNLOAD")

# download all files to data-raw/ ----
mapply(function(i, j)
  download.file(i, destfile = file.path("data-raw", j))
  , chicago.data
  , names(chicago.data)
  , SIMPLIFY = FALSE)

# end of script #


data-raw/02_export_raw_data.R
#
# Author:   Cristian E. Nuno
# Purpose:  Export raw data sets as compressed .rda files
# Date:     March 14, 2019
#

# load necessary packages ----
# note: as the world moves forward with the `sf` package
#       I will save internal data as both sf and spdf objects
library(sf)
library(rgdal)

# load all objects ----
community_areas_spdf <- readOGR(dsn = "data-raw/chicago_community_area_boundaries.geojson"
                                , stringsAsFactors = FALSE)
community_areas_sf   <- read_sf("data-raw/chicago_community_area_boundaries.geojson")

city_boundary_spdf   <- readOGR(dsn = "data-raw/chicago_city_boundary.geojson"
                                , stringsAsFactors = FALSE)

city_boundary_sf     <- read_sf("data-raw/chicago_city_boundary.geojson")

cps_sy1819           <- read.csv("data-raw/chicago_public_schools_sy1819.csv"
                                 , stringsAsFactors = FALSE
                                 , na.strings = "")
census_tracts_spdf   <- readOGR(dsn = "data-raw/chicago_2010_census_tracts.geojson"
                                , stringsAsFactors = FALSE)
census_tracts_sf     <- read_sf("data-raw/chicago_2010_census_tracts.geojson")

# clean community_areas_spdf ----
# note: all other columns but community & area number are either zero or irrelevant
community_areas_spdf <-
  community_areas_spdf[, c("community", "area_numbe")]

# clean community_areas_sf -----
# note: keeping only the community, area number and geometry columns
community_areas_sf <-
  community_areas_sf[, c("community", "area_numbe", "geometry")]

# clean cps_sy1819 ----
names(cps_sy1819) <- tolower(names(cps_sy1819))

# store columns where values are 'true' or 'false'
tf.char.columns <-
  c("is_high_school", "is_middle_school", "is_elementary_school"
    , "is_pre_school", "attendance_boundaries", "dress_code"
    , "bilingual_services", "refugee_services", "title_1_eligible"
    , "significantly_modified", "hard_of_hearing", "visual_impairments"
    , "is_gocps_participant", "is_gocps_prek", "is_gocps_elementary"
    , "is_gocps_high_school")

# replace character values with logical values
cps_sy1819[, tf.char.columns] <-
  lapply(X = cps_sy1819[, tf.char.columns], FUN = as.logical)

# store columns where values are 'Y'
yes.char.columns <-
  c("preschool_inclusive", "preschool_instructional")

# replace character values with logical values
cps_sy1819[, yes.char.columns] <-
  lapply(X = cps_sy1819[, yes.char.columns]
         , FUN = function(i) ifelse(i == "Y", TRUE, i))

# store all non-administrator contacts & titles
non.administrator.contacts.titles <-
  names(cps_sy1819)[grepl("contact", names(cps_sy1819))]

# only keep columns that aren't in non.administrator.contacts.titles vector
cps_sy1819 <-
  cps_sy1819[, !names(cps_sy1819) %in% non.administrator.contacts.titles]

# drop the act columns since every record is NA
act <-
  names(cps_sy1819)[grepl("_act", names(cps_sy1819))]

cps_sy1819 <-
  cps_sy1819[, !names(cps_sy1819) %in% act]

# drop the closed_for_enrollment_date since every record is NA
closed <-
  "closed_for_enrollment_date"

cps_sy1819 <-
  cps_sy1819[, !names(cps_sy1819) %in% closed]

# Englewood STEM HS has the incorrect
# graduation rate mean and college enrollment rate mean rates
cps_sy1819$graduation_rate_mean <-
  ifelse(cps_sy1819$graduation_rate_mean == 77.5
         , 78.2
         , cps_sy1819$graduation_rate_mean)

cps_sy1819$college_enrollment_rate_mean <-
  ifelse(cps_sy1819$college_enrollment_rate_mean == 59.8
         , 68.2
         , cps_sy1819$college_enrollment_rate_mean)

# cast overall_rating as a factor so
# that school quality rating policy (SQRP) ranking is upheld
cps_sy1819$overall_rating <-
  factor(cps_sy1819$overall_rating
         , levels = c("Level 1+"
                      , "Level 1"
                      , "Level 2+"
                      , "Level 2"
                      , "Level 3"
                      , "Inability to Rate"))

# for no reason, the data is missing the school type field from SY1617
# and has replaced it with the lenghtly classification_description column
# thankfully, they used the exact same text as before so
# the values can be mapped perfectly
cps_sy1617 <-
  readRDS(gzcon(url("https://github.com/cenuno/Learning_R/raw/master/write_data/clean_rating_cps_sy1617_school_profiles.rds")))

classification.type <-
  as.data.frame(table(cps_sy1617$School_Type
                      , cps_sy1617$Classification_Description))

names(classification.type) <-
  c("classification_type"
    , "classification_description_sy1617"
    , "n")

classification.type <-
  classification.type[classification.type$n > 0, ]

# add the classification_type to cps_sy1819
cps_sy1819 <-
  merge(x = cps_sy1819
        , y = classification.type
        , by.x = "classification_description"
        , by.y = "classification_description_sy1617"
        , all.x = TRUE)

# remove the 'n' column so it's no longer relevant
cps_sy1819$n <- NULL

# note: base::merge() rearranged the columns so I'm manually arranging them here
cps_sy1819 <-
  cps_sy1819[c(2:73, 81, 1, 74:80)]

# transform open enrollment date to Date object
cps_sy1819$open_for_enrollment_date <-
  as.Date(cps_sy1819$open_for_enrollment_date, format = "%m/%d/%Y")

# remove the closed_for_enrollment_date since it's all NA values
cps_sy1819$closed_for_enrollment_date <- NULL

# cast school_id as a character vector
cps_sy1819$school_id <- as.character(cps_sy1819$school_id)

# ensure a few text columns are encoded as UTF-8
for (i in c("long_name", "summary", "administrator"
            , "statistics_description", "demographic_description"
            , "rating_statement", "classification_description")) {
  Encoding(cps_sy1819[[i]]) <- "UTF-8"
}

# clean census tracts ----
census_tracts_sf$notes <-
  # replace "" with NA values in the notes column
  ifelse(census_tracts_sf$notes == "", NA, census_tracts_sf$notes)

census_tracts_spdf@data$notes <-
  # replace "" with NA values in the notes column
  ifelse(census_tracts_spdf@data$notes == ""
         , NA
         , census_tracts_spdf@data$notes)

# export all objects as .rda files in data/ -----
save(community_areas_spdf
     , file = "data/community_areas_spdf.rda"
     , compress = "xz")
save(community_areas_sf
     , file = "data/community_areas_sf.rda"
     , compress = "xz")
save(city_boundary_spdf
     , file = "data/city_boundary_spdf.rda"
     , compress = "xz")
save(city_boundary_sf
     , file = "data/city_boundary_sf.rda"
     , compress = "xz")
save(cps_sy1819
     , file = "data/cps_sy1819.rda"
     , compress = "bzip2")
save(census_tracts_spdf
     , file = "data/census_tracts_spdf.rda"
     , compress = "xz")
save(census_tracts_sf
     , file = "data/census_tracts_sf.rda"
     , compress = "xz")

# end of script #


R/02_label_points_within_polygons.R
#
# Author:   Cristian E. Nuno
# Purpose:  Label points within polygons
# Date:     March 14, 2019
#

# create LabelPointsWithinPolygons() function -----
#' @export
LabelPointsWithinPolygons <- function(lng
                                      , lat
                                      , polygon.boundaries) {
  # Labels the polygon each coordinate pair lies in
  #
  # Inputs:
  # - lng: a numeric vector of non-NA longitudinal points
  # - lat: a numeric vector of non-NA latiduinal points
  # - polygon.boundaries: either a coordinate pair matrix or
  #                       a named list of coordinate pair matrices

  # Output:
  # if polygon.boundaries is a coordinate pair matrix,
  # a logical vector will be returned identifying
  # those points which lie in the polygon
  #
  # else, a character vector will be returned
  # identifying the polygon each coordinate pair lies in

  # ensure lng is a numeric vector ----
  if (is.numeric(lng) == FALSE) {
    stop("lng must be a numeric vector.")
  }

  # ensure lat is a numeric vector ----
  if (is.numeric(lat) == FALSE) {
    stop("lat must be a numeric vector.")
  }

  # ensure lng and lat are of equal length ----
  if (length(lng) != length(lat)) {
    stop("lng and lat must be of equal length.")
  }

  # ensure lng contains no NA values ---
  if (any(is.na(lng))) {
    stop("Filter lng and lat so that neither contain any NA values.")
  }

  # ensure lat contains no NA values ----
  if (any(is.na(lat))) {
    stop("Filter lng and lat so that neither contain any NA values.")
  }

  # ensure class of polygon.boundaries is a matrix or a list of matrices ----
  if (class(polygon.boundaries) != "matrix") {
    if (!is.list(polygon.boundaries)) {
      stop("polygon.boundaries must either be of class matrix or a list of matrices.\n See ?GetPolygonBoundaries for more information.")
    }
  }

  if (class(polygon.boundaries) != "list") {
    if (!is.matrix(polygon.boundaries)) {
      stop("polygon.boundaries must either be of class matrix or a list of matrices. See ?GetPolygonBoundaries for more information.")
    }
  }

  # if polygon.boundaries is a list, ensure that each matrix is labeled ----
  if (is.list(polygon.boundaries) &
      is.null(names(polygon.boundaries))) {
    stop("Please label the matrices within polygon.boundaries. See ?names for more details.")
  }

  # splancs::inpip() requires the longitude be the first column and latitude the second ----
  # note: df$index will be ignored inside splancs::inpip() as it only subsets the first two columns of pts
  df <-
    data.frame(x = lng, y = lat, index = 1:length(lng))

  # return a logical vector if any points lie within the 1 coordinate pair matrix ----
  if (is.matrix(polygon.boundaries)) {
    return(df$index %in%
             splancs::inpip(pts = df
                            , poly = polygon.boundaries
                            , bound = NULL))
  } else {
    # identify which points lie inside each coordinate pair matrix ----
    temp <-
      lapply(X = polygon.boundaries
             , FUN = function(i)
               splancs::inpip(pts = df
                              , poly = i
                              , bound = NULL))

    # name elements in temp
    names(temp) <- names(polygon.boundaries)

    # collapse the list of indices and polygon labels into one data frame
    # note: instances where no points were found in a polygon will be dropped
    temp <- utils::stack(temp)

    # rename columns
    names(temp)[names(temp) == "values"] <- "index"
    names(temp)[names(temp) == "ind"] <- "label"

    # transfrom the label column from a factor to a character
    temp$label <- as.character(temp$label)

    # test for points that don't exist in any of the polygons ----
    if (nrow(df) > nrow(temp)) {

      no.label.points <-
        data.frame(index = df$index[which(!df$index %in% temp$index)]
          , label = NA)

      temp <- rbind(temp, no.label.points)
    }

    # ensure labels are in the order of the input coordinate pairs ----
    temp <- temp[order(temp$index), ]

    return(temp$label)
  }

} # end of LabelPointsWithinPolygons() function

# end of script #


data-raw/03_create_logo.R
#
# Author:   Cristian E. Nuno
# Purpose:  Create logo for pointdexter
# Date:     January 12, 2019
#

# load necessary package ----
library(sp)

# load necessary data ----
load(file = "data/community_areas_spdf.rda")
load(file = "data/cps_sy1819.rda")

# store max and min latitudinal coordinates ---
north.lat <- 41.87249
south.lat <- 41.81995

# filter cps schools to those that meet the condition ----
df <-
  cps_sy1819[cps_sy1819$is_high_school &
               c(cps_sy1819$overall_rating == "Level 1+" |
                   cps_sy1819$overall_rating == "Level 1") &
               (cps_sy1819$school_latitude >= north.lat |
                  cps_sy1819$school_latitude <= south.lat), ]

# define new mapping for avenir font ----
quartzFonts(avenir = c("Avenir Book", "Avenir Black", "Avenir Book Oblique",
                       "Avenir Black Oblique"))

# plot all 77 community areas and export results as png ----

# open up plot device
png("man/figures/base_logo.png"
    , width = 6.777778
    , height = 5.833333
    , units = "in"
    , res = 600)

# customize plotting parameters
par(family = "avenir"
    , mar = c(0.75, 0.75, 0, 1.15))

# plot community areas
plot(community_areas_spdf
     , border = "#3E3D3C")

# add text in the middle of the plot
text(x = -87.72
     , y = 41.84
     , labels = "pointdexter"
     , cex = 5.7
     , col = "#D44500"
     , font = 2)

# add copyright text to plot
text(x = -87.83062
     , y = 41.66055
     , labels = "http://cenuno.github.io/pointdexter/"
     , cex = 0.75
     , font = 2
     , col = "#D44500")

# add school points
points(x = df$school_longitude
       , y = df$school_latitude
       , pch = 19
       , col = "dodgerblue4"
       , cex = 1.45
       , adj = 0.5)

dev.off()

# note: -----
# base_logo.png is then taken into Microsoft Word for cropping
# it into the shape of a hexagon.
#
# the modified base_logo.png is then saved as logo.png

# end of script #
